Generate reference images from LaTeX

gagern opened the pull request:
In https://github.com/Khan/KaTeX/pull/156#issuecomment-114572831, @kevinb7 wrote that he'd like to see TeX renderings of our examples before proceeding. Here they are. The docker may still be improved upon, I guess, in terms of size and similar, but it works. I also added a small tool to compute visual diffs, but haven't committed the results of this. One thing I notice is that in TeX, `\llap` and `\rlap` apparently take an argument in text mode.

kevinb7 commented:
This is awesome!  I was trying to the square root rendering again and started comparing against LaTeX manually and noticed that the amount space under the hrule differs quite a bit between KaTeX and LaTeX.  I'll post some screenshots later, but I was wondering if you noticed any differences for the square root test case?

gagern commented:
Which test case, `Sqrt` or `SqrtRoot`? I noticed considerable differences for both:

![sqrt](https://cloud.githubusercontent.com/assets/190759/8358181/3a3f3496-1b5f-11e5-9816-817740ec2b59.png)

![sqrtroot](https://cloud.githubusercontent.com/assets/190759/8358205/406d2580-1b5f-11e5-9c5d-382e3c9bcb25.png)

The horizontal spacing in the second one seems to be largely affected by the proportions of the digits in the root. I guess this is because LaTeX has a separate font for smaller sizes, while KaTeX simply scales the 10pt font down, right?

gagern commented:
> I also added a small tool to compute visual diffs, but haven't committed the results of this. 

I just committed and pushed these diffs to [a separate branch](https://github.com/gagern/KaTeX/tree/diffs/test/screenshotter/diff). I'm not sure if you want this in repo. I don't think I'd want it there, since they are easily reproduced as long as you have imagemagick around. But it might be convenient to have a look now.

spicyj commented:
> I guess this is because LaTeX has a separate font for smaller sizes, while KaTeX simply scales the 10pt font down, right?

This certainly causes some deviations. We tend to just ignore it.

kevinb7 commented:
@gagern I was look at ```Sqrt``` thanks for posting ```SqrtRoot``` too.  Thanks for posting all of the screenshots.  It's nice to know where were at in terms of LaTeX compliance for current features.  I agree we can just regenerate these as needed.  When we improve any of the renderings we'll update the screenshotter files.  I'm not sure we need to store the -pdflatex.png images either.

gagern commented:
> I'm not sure we need to store the -pdflatex.png images either.

I'm not sure how different those are from the `*-firefox.png` files. Both can be recreated exactly from the sources with the help of a docker, which requires a working docker setup and considerable amounts of HDD space for the relevant images. Is there any reason to keep the KaTeX snapshots but not their reference images? Is it for the sake of documenting changes to the code base which affect screenshots?

kevinb7 commented:
The difference is that we use ```*-firefox.png``` as reference files to identify unintended changes in the rendering.  We do this by running the snapshotter and seeing if there are any differences in the files according to git.  This process should be automated so that we can fail the build if there are difference, but right now it's a manual process.

If there are, either the code should be fixed so the rendering matches or the reference file should be updated because we fixed/improved the rendering.

On the other hand ```*-pdflatex.png``` images aren't being compared against old version of themselves so we don't need to have them in the repo.

kevinb7 commented:
I was looking through some of the ```-pdflatex.png``` renderings and some of them are clipped.  Do you know why this happening?

gagern commented:
> some of them are clipped.

That's because I'm aligning the images to maximize overlap, but using the canvas from the Firefox images for reference. So if after that overlap-maximizing alignment, the TeX version extends beyond the margin of the Firefox screenshot, it will get clipped. I fear that without that alignment step, many of the images would have some constant offset over large parts of the canvas, making relative changes harder to see.

kevinb7 commented:
@gagern I would like merge this in.  I'll try to do a better review next week.  Could you add some comments in texcmp.js to describe what's going on?  I assume that you're using the fft to align images.  

I was looking at https://github.com/gagern/KaTeX/blob/diffs/test/screenshotter/diff/SizingBaseline.png and realized that the reason why it's so off is because LaTeX doesn't support sizing inside math.  And with respect to https://github.com/gagern/KaTeX/blob/diffs/test/screenshotter/diff/NestedFractions.png I see that it's lining things up with the bigger fractions and b/c our spacing is wrong the LaTeX screenshot gets clipped.

Could you make it so that the bounding box of the diffs is the union of the bounding boxes of the images being compared?

Also, if there are things that we know are not going to yield invalid comparisons, i.e. sizing commands inside math, can skip those examples?

kevinb7 commented:
I think the reason why the exponents are a bit off is b/c LaTeX actually stretches them horizontally a bit (or uses different glyphs).

spicyj commented:
It uses different glyphs. Same as https://github.com/Khan/KaTeX/pull/268#issuecomment-115293652.

gagern commented:
> I would like merge this in.

Without the actual images, right? Will rewrite when I find the time.

> Could you make it so that the bounding box of the diffs is the union of the bounding boxes of the images being compared?

The way I implemented things now, I rasterize the PDF to a temporary file, then align that to produce the `*-pdflatex.png`. So at that point, the clipping has already occurred. The reason I do it this way is because I sometimes want to flip between two renderings, and see what changes. If I move the aligning from the code running inside the docker to the diff, that will no longer be possible that easily. What do you think?

> can skip those examples?

Sure, we can add those to the blacklist.

spicyj commented:
Can we just add a little padding to the katex screenshots so that the pdflatex ones don't later get cut off?

gagern commented:
Some of them are nearly exhausting their 1024×768 screenshot size. E.g. `OpLimits` and `Spacing`. If we add margin or padding to the `test.html`, these might start scrolling unless we increase the screenshoit size. Is that desirable?

kevinb7 commented:
> Without the actual images, right? Will rewrite when I find the time.

That's right.  Whenever you have time.

I think @spicyj meant expanding the bounds of the KaTeX images.  After alignment we should know the difference in bounds and be able to pad either image accordingly.  We probably want to keep the existing reference screenshots as is so this would be another set of temp images.  These images could be used for flipping between and generate visualize diffs without clipping any of the renderings.

The clipping doesn't need to be perfect for the images to be useful as indicated by the number of new issues opened due to rendering differences.  I think we should merge it and improve the clipping later.

spicyj commented:
Well, whichever works. We could probably reduce the font size and be okay if we're running out of room.

gagern commented:
OK, rewrote things, incorporating your requests. The rasterized images are now not artificially aligned, but use the dimensions of the underlying document. The alignment is computed but used only for the diffs. So these are indeed rendered to the union of both pictures, then trimmed to the area that's actually being used. I added some comments about what's happening in [`texcmp.js`](https://github.com/gagern/KaTeX/blob/texcmp/dockers/texcmp/texcmp.js). And I omitted the resulting images, although I added them [on another branch](https://github.com/gagern/KaTeX/tree/diffs/test/screenshotter/tex). Hope this helps.

I also implemented some fake math sizing commands (although they still [don't work with fractions](http://tex.stackexchange.com/q/252815/16923)), a `\blue` command and something to handle display mode, pre and post text chunks. This makes the images resemble the ones from Firefox much better. The blacklist now contains only three items, namely `Colors`, `DeepFontSizing` and `KaTeX`.

kevinb7 commented:
@gagern Awesome, thanks!  I'll try to review this evening.

xymostech commented on dockers/texcmp/Dockerfile:
```diff
@@ -0,0 +1,5 @@
+FROM ubuntu:15.04
+MAINTAINER Martin von Gagern <gagern@ma.tum.de>
+RUN sed -i 's/^\(deb.*updates\)/#\1/' /etc/apt/sources.list && apt-get update
```
What does this do?

gagern replied:
Disable normal updates, so we don't receive any bug fix releases or similar, just security relevant fixes. The benefit is that we don't have to pin the versions of the packages we install (which are quite numerous, with all the indirect deps pulled in), and still rely on the same versions getting installed in subsequent runs unless something drastic forces a security update to one of the packages. It's one more step towards better reproducibility.

kevinb7 replied:
Can you add a comment here to that effect?

xymostech commented on dockers/texcmp/texcmp.js:
```diff
@@ -0,0 +1,244 @@
+]).spread(function(data) {
+    template = data;
+    // dirs have been created, template has been read, now rasterize.
+    return Q.all(todo.map(processTestCase));
```
If you do something like `processTestCase.bind(null, data)` and add an argument to `processTestCase`, you won't have to make the global `template` variable.

gagern replied:
True. I considered the module level variable to be easier to understand and maintain than some `bind` magic.

gagern replied:
At some point in the past, `template` was the result of a call to `readFileSync`, so this is partly historic. But we might get back to something like this one day, for some reason I can#t imagine right now, and then the global variable is more flexible because we don't have to care where it gets set, as long as it does get set by the time processTestCase is invoked.

kevinb7 commented on dockers/texcmp/README.md:
```diff
@@ -0,0 +1,73 @@
+Note that the files and directories created inside the docker will be
+owned by root, so you might have trouble deleting them later on.  Be
+sure you can obtain superuser permissions on your computer or know
+someone who can, just to be safe.
```
These are files on the host system created by docker, correct?

gagern replied:
Right. I wonder whether we should be streaming out those files via a TCP connection instead of a `--volume` mount. Or pre-create the files outside. Or anything else like this. But I see that as a separate pull request, once the core functionality is in place.

kevinb7 replied:
Fair enough.  Thanks for calling it out in the docs.  Pre-creating the files seems like it would be pretty easy to implement.

kevinb7 commented on dockers/texcmp/README.md:
```diff
@@ -0,0 +1,73 @@
+### Native environment
+
+If you want to avoid the overhead of creating a docker container, or
+the even large roverhead of setting up docker and creating the initial
```
large roverhead => larger overhead

kevinb7 commented on dockers/texcmp/README.md:
```diff
@@ -0,0 +1,73 @@
+Note that this approach will use `/tmp/texcmp` as a temporary directory.
+The use of a single directory name here can lead to conflicts if
+multiple developers on the same machine try to use that directory.
+
```
It would be good to remind people that the results from this method will differ from machine to machine.

kevinb7 commented on dockers/texcmp/README.md:
```diff
@@ -0,0 +1,73 @@
+## Generated files
+
+After running either of the above commands, you will find two
+(possibly new) subdirectories.
```
It would be good to list the subdirectories here.

kevinb7 commented on dockers/texcmp/README.md:
```diff
@@ -0,0 +1,73 @@
+coloring just described.  The pictures will be aligned in such a way
+as to maximize the overlap between the two versions (i.e. the amount
+of black output).  The result will then be trimmed so it can easily be
+pasted into bug reports.
```
Nice!

kevinb7 commented on dockers/texcmp/README.md:
```diff
@@ -0,0 +1,73 @@
+Both `texcmp.sh` and `texcmp.js` will accept the names of test cases
+on the command line.  This can be useful if one particular test case
+is affected by current development, so that the effects on it can be
+seen more quickly.
```
It would be good to have an example here.

kevinb7 commented on dockers/texcmp/texcmp.js:
```diff
@@ -0,0 +1,244 @@
+    pngparse.parseStream(stream, parsed);
+    return deferred.promise;
+
+    function parsed(err, image) {
```
Could you extract the body of the function (everything below the ```if (err) ...``` check and put it in a function called ```calcAlignment``` (or something like that)?

kevinb7 commented on dockers/texcmp/texcmp.js:
```diff
@@ -0,0 +1,244 @@
+}
+
+// Read given file, and apply FFT transformation to it.
+function fftPNG(file) {
```
Can you break this up into separate functions, one for reading the PNG and one for computing the FFT on it?

gagern replied:
I can. In that case, the promise would return the parsed image, and almost everything in `parsed` would be a separate function.

kevinb7 replied:
That's okay.  Could you instead inline ```parsed``` so that we're not relying on hoisting for this work?  It's weird to have code appear after a ```return```.

kevinb7 commented on dockers/texcmp/texcmp.js:
```diff
@@ -0,0 +1,244 @@
+        // Step 6c: find position where the (squared) absolute value is maximal
+        var ox = 0;
+        var oy = 0;
+        var ov = -1;
```
```ox``` == ```overlapX```?  What's ```ov``` and why is it ```-1```?

kevinb7 replied:
```offsetX```?

kevinb7 replied:
`offsetX`?

gagern replied:
`ox` and `oy` stand for offset, yes. `ov` is the value associated with the best offset, or to be more precise, the squared absolute value. Since that will never be negative, initializing it to a negative number ensures that the first actual pixel will overwrite that value.

kevinb7 replied:
Maybe ```ov``` => ```maxSquaredNorm``` and ```sq``` => ```squaredNorm```?

kevinb7 commented on dockers/texcmp/texcmp.js:
```diff
@@ -0,0 +1,244 @@
+}
+
+// Create a new matrix of preconfigured dimensions, initialized to zero
+function newMatrix() {
```
Can you switch this to ```createMatrix```? ```newMatrix()``` is too close to ```new Matrix()```.

gagern replied:
To me, it feels conceptually close, so this was intentional. But I have no strong opinion there, `createMatrix` is fine with me, too.

kevinb7 commented on dockers/texcmp/texcmp.js:
```diff
@@ -0,0 +1,244 @@
+        var bx = Math.max(ox, 0);
+        var by = Math.max(oy, 0);
+        var lx = Math.max(-ox, 0);
+        var ly = Math.max(-oy, 0);
```
Can you make the labels a bit more verbose?

gagern replied:
Can I add comments instead? I'd prefer to keep the `-extent` arguments in one line.

kevinb7 replied:
that works too

kevinb7 commented on dockers/texcmp/texcmp.js:
```diff
@@ -0,0 +1,244 @@
+            "-extent", uw + "x" + uh + "-" + bx + "-" + by,
+            ")",
+            "(", "-clone", "0-1", "-compose", "darken", "-composite", ")",
+            "-channel", "RGB", "-combine", "-trim", diffFile
```
template strings would really help here ^_^

kevinb7 replied:
why the parens?

gagern replied:
The parenthesized parts describe a sequence of operations which is applied to some input and the result is then used as the input on the outer scope. So the first parenthesis reads the LaTeX file, converts that to grayscale and pads it to the union. The second does the same for the browser screenshot. The third takes the first two and takes the minimal gray value for each pixel. These three then form the red, green and blue channel of the final image. See http://www.imagemagick.org/script/command-line-processing.php#stack for details.

kevinb7 replied:
This was really helpful, can you add it as a comment?

kevinb7 commented on dockers/texcmp/texcmp.js:
```diff
@@ -0,0 +1,244 @@
+var template;
+
+Q.all([
+    Q.nfcall(fs.readFile, path.join(ssDir, "test.tex"), "utf-8"),
```
you might want to look at denodeify instead of using ```Q.nfcall```

gagern replied:
My personal policy is that I go with `nfcall` if I use it in only one or perhaps two source lines, and `denodeify` if I call it in more different source lines, or as part of a tight loop. Can I go with that here, or should I `denodeify` everything for the sake of consistency?

kevinb7 replied:
Let's go with ```denodeify``` for consistency.

kevinb7 commented on dockers/texcmp/texcmp.js:
```diff
@@ -0,0 +1,244 @@
+    return Q.nfcall(fs.mkdir, dir).catch(function(err) {
+        if (err.code !== "EEXIST")
+            throw err;
+    });
```
Can you use ```mkdirp``` instead?  https://github.com/substack/node-mkdirp

gagern replied:
Sure. I don't need to create parent directories, but if you prefer extra dependencies, I can do that.

kevinb7 replied:
As long as we keep katex.js free from extra dependencies. ;)  This is a very minor detail so we can leave it as is.

kevinb7 commented on dockers/texcmp/texcmp.js:
```diff
@@ -0,0 +1,244 @@
+            console.error(stdout + stderr);
+            err.stdout = stdout;
+            err.stderr = stderr;
+            deferred.reject(err);
```
denodeify would help here too

gagern replied:
I don't think I can use `denodeify` to get access to the `stderr` content. If the promise gets rejected, all information from the accepting path is lost, afaik.

kevinb7 replied:
Good point.

gagern commented:
If you are happy with all the new changes, I'd squash everything into a single commit before merging it.

kevinb7 commented:
@gagern the new changes look great!  Squash it and I'll merge it.

kevinb7 commented:
Thanks for the pull request.

gagern commented:
Good to have this in. Thanks!
